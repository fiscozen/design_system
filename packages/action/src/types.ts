import { IconVariant } from "@fiscozen/icons";
import { FzLinkProps } from "@fiscozen/link";
import { RouteLocation } from "vue-router";

// Environment variants
export type FzActionEnvironment = "backoffice" | "frontoffice";

// Layout variants
export type FzActionVariant = "textLeft" | "textCenter" | "onlyIcon";

// Common props shared by both link and button actions
export interface FzActionCommonProps {
  /**
   * Name of the fontawesome icon
   */
  iconName?: string;
  /**
   * Variant of the fontawesome icon
   */
  iconVariant?: IconVariant;
  /**
   * Alternative prop to default label slot
   */
  label?: string;
  /**
   * Sub-label text (smaller text below the main label)
   */
  subLabel?: string;
  /**
   * Disables the action
   */
  disabled?: boolean;
  /**
   * Environment variant (backoffice/frontoffice)
   */
  environment?: FzActionEnvironment;
  /**
   * Layout variant (textLeft/textCenter/onlyIcon)
   */
  variant?: FzActionVariant;
  /**
   * Whether text should be truncated
   */
  isTextTruncated?: boolean;
}

// Props specific to link actions (from FzLink)

// Discriminated union for action props
export type FzActionProps =
  | (FzActionCommonProps &
      FzLinkProps & {
        type: "link";
        meta: CustomRouteLocation;
      })
  | (FzActionCommonProps & {
      type: "action";
    });

// Helper type for Vue Router route locations
type PartialExcept<T, K extends keyof T> = Pick<Required<T>, K> & Partial<T>;

export type CustomRouteLocation = PartialExcept<RouteLocation, "path" | "name">;
