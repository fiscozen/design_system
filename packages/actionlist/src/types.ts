import { FzRouterNavlinkProps, FzNavlinkProps } from '@fiscozen/navlink'

type ActionlistItem<T = void> =
  | (FzRouterNavlinkProps & { type: 'link' })
  | (FzNavlinkProps<T> & { type: 'button' })

interface FzActionlistProps<T = void> {
  label?: string;
  items: ActionlistItem<T>[];
  listClass?: string;
}

export { ActionlistItem, FzActionlistProps }
