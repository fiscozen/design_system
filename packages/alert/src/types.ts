import { RouteLocationRaw } from 'vue-router'

type AlertProps = {
  /**
   * Type of the topbar which dictates the appearance
   */
  type: 'info' | 'error' | 'danger' | 'warning' | 'success'
  /**
   * Style which dictates the action rendered
   */
  alertStyle?: 'default' | 'collapsable' | 'simple'
  /**
   * Whether to render the button action
   */
  showButtonAction?: boolean
  /**
   * Button action label
   */
  buttonActionLabel?: string
  /**
   * Button action tooltip
   */
  buttonActionTooltip?: string
  /**
   * Whether to render the link action
   */
  showLinkAction?: boolean
  /**
   * Link action label
   */
  linkActionLabel?: string
  /**
   * Action link location (applies only if showLinkAction is true)
   */
  linkActionLocation?: RouteLocationRaw
  /**
   * Size of the alert (applies only if style is simple)
   */
  size?: 'sm' | 'md' | 'lg'
  /**
   * Title of the alert (applies only if style is not simple)
   */
  title?: string
  /**
   * Default state of the alert (open/closed)
   */
  defaultOpen?: boolean
}

export { AlertProps }
