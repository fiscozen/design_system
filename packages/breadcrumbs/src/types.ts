import { RouteLocation, RouterLinkProps } from 'vue-router'

export interface Breadcrumb<T = void> {
  id: string
  label: string
  metadata: T extends {} ? T : undefined
}

type PartialExcept<T, K extends keyof T> = Pick<Required<T>, K> & Partial<T>

export type CustomRouteLocation = PartialExcept<RouteLocation, 'path' | 'name'>
