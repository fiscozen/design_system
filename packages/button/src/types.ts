/**
 * Type definitions for the Fiscozen Button component library.
 * 
 * @module @fiscozen/button/types
 */

import type { IconVariant, IconSize } from '@fiscozen/icons'

export type CommonButtonVariant = 'primary' | 'secondary' | 'invisible'
export type IconButtonVariant = CommonButtonVariant | 'notification'
export type ButtonVariant = CommonButtonVariant | 'danger' | 'success'
export type ButtonSize = 'xs' | 'sm' | 'md' | 'lg'
export type ButtonEnvironment = 'backoffice' | 'frontoffice'

/**
 * Type for size to environment mapping record
 */
export type SizeToEnvironmentMap = Record<ButtonSize, ButtonEnvironment>

/**
 * Type for button size to icon size mapping record
 */
export type IconSizeMap = Record<ButtonSize, IconSize>

/**
 * Props for the FzButton component.
 * 
 * A comprehensive button component supporting five visual variants and two environment-based sizes.
 * Features include icon integration, multiple interactive states, and full accessibility support.
 * 
 * @example
 * <FzButton label="Click me" variant="primary" />
 */
export interface FzButtonProps {
  /**
   * Text label displayed on the button. Overridden by default slot if provided.
   */
  label?: string
  /**
   * @deprecated Use FzTooltip component to wrap the button instead.
   * A runtime warning will be displayed when this prop is used.
   */
  tooltip?: string
  /**
   * Visual style variant determining colors and interactive states
   * @default 'primary'
   */
  variant?: ButtonVariant
  /**
   * Button size affecting height, padding, and text size
   * @deprecated Use the 'environment' prop instead. This prop will be removed in a future version.
   */
  size?: ButtonSize
  /**
   * Environment determining button size
   * @default 'frontoffice'
   */
  environment?: ButtonEnvironment
  /**
   * Disables interaction and applies disabled styling
   * @default false
   */
  disabled?: boolean
  /**
   * FontAwesome icon name (e.g., 'bell', 'save'). Search: https://fontawesome.com/v6/icons
   */
  iconName?: string
  /**
   * FontAwesome icon variant (e.g., 'fas', 'far', 'fal')
   */
  iconVariant?: IconVariant
  /**
   * Icon position relative to label
   * @default 'before'
   */
  iconPosition?: 'before' | 'after'
  /**
   * Additional CSS classes for label container. Merged with 'truncate' unless overrideContainerClass is true.
   */
  containerClass?: string
  /**
   * Replaces default container classes with containerClass instead of merging
   * @default false
   */
  overrideContainerClass?: boolean
}

/**
 * Props for the FzIconButton component.
 * 
 * Icon-only button variant with notification badge support.
 * 
 * @example
 * <FzIconButton iconName="bell" variant="notification" />
 */
export interface FzIconButtonProps {
  /**
   * The tooltip of the button will be shown on hover
   */
  tooltip?: string
  /**
   * primary or secondary button
   * @default 'primary'
   */
  variant?: IconButtonVariant
  /**
   * size of the button
   * @default 'md'
   */
  size?: ButtonSize
  /**
   * whether action is enabled or not
   * @default false
   */
  disabled?: boolean
  /**
   * Icon to be displayed. Use fontawesome search here https://fontawesome.com/v6/icons
   * Mind that not all of the icons and variants might be available.
   */
  iconName: string
  /**
   * Fontawesome icon variant: solid, regular, light, thin. Sharp subvariants are available as well
   * @default 'far'
   */
  iconVariant?: IconVariant
  /**
   * Icon size
   */
  iconSize?: IconSize
  /**
   * Positioning of the icon
   * @default 'before'
   */
  iconPosition?: 'before' | 'after'
  /**
   * Additional class for the icon
   * @default ''
   */
  iconClass?: string
}

/**
 * Props for the FzButtonGroup component.
 * 
 * Container component for grouping multiple buttons with consistent spacing.
 * 
 * @example
 * <FzButtonGroup horizontal :gap="true">
 *   <FzButton label="Save" />
 *   <FzButton label="Cancel" />
 * </FzButtonGroup>
 */
export interface FzButtonGroupProps {
  /**
   * The direction of the button group. If true, buttons will be displayed horizontally, otherwise vertically
   * @default true
   */
  horizontal?: boolean
  /**
   * Whether to add a gap between buttons
   */
  gap?: boolean
  /**
   * The gap size between buttons
   * @default 'md'
   */
  size?: ButtonSize
}
