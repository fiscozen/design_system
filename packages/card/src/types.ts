export type FzCardProps = {
  title: string;
  color?: FzCardColor;
  primaryAction?: FzCardButton;
  secondaryAction?: FzCardButton;
  tertiaryAction?: FzCardIconButton;
  contentClass?: string;
};

type FzCardButton = {
  label: string;
  callback: () => void;
};

type FzCardIconButton = {
  icon: string;
  callback: () => void;
};

export enum FzCardColor {
  Blue = "blue",
  Orange = "orange",
  Purple = "purple",
}
