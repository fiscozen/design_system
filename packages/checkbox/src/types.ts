import { FzTooltipProps } from "@fiscozen/tooltip";

export type FzCheckboxProps = {
  /**
   * The label of the checkbox
   */
  label: string;
  /**
   * The value of the checkbox. If not provided, the value will be the label
   */
  value?: string | number | boolean;
  /**
   * The size of the checkbox
   */
  size?: "sm" | "md";
  /**
   * if the checkbox is indeterminate
   */
  indeterminate?: boolean;
  /**
   * If true, the checkbox will be emphasized
   */
  emphasis?: boolean;
  /**
   * If true, the checkbox will be disabled
   */
  disabled?: boolean;
  /**
   * If true, the checkbox will be in an error state
   */
  error?: boolean;
  /**
   * If the checkbox is required
   */
  required?: boolean;
  /**
   * If the checkbox is standalone
   */
  standalone?: boolean;
  /**
   * Tooltip props for the checkbox
   */
  tooltip?: FzTooltipProps;
  /**
   * ARIA owns attribute for hierarchical checkboxes
   */
  ariaOwns?: string;
  /**
   * Custom ID for the checkbox (if not provided, a random ID will be generated)
   */
  checkboxId?: string;
};

export type FzCheckboxGroupProps = {
  /**
   * The label of the checkbox group
   */
  label: string;
  /**
   * The size of the checkbox
   */
  size: "sm" | "md";
  /**
   * The checkbox to render
   */
  options: ParentCheckbox[];
  /**
   * If true, the checkbox will be emphasized
   */
  emphasis?: boolean;
  /**
   * If true, the checkbox will be disabled
   */
  disabled?: boolean;
  /**
   * If true, the checkbox will be in an error state
   */
  error?: boolean;
  /**
   * If the checkbox group is required
   */
  required?: boolean;
  /**
   * If true, the checkbox group will be horizontal
   */
  horizontal?: boolean;
};
export type ParentCheckbox = ChildCheckbox & { children?: ChildCheckbox[] };
export type ChildCheckbox = Omit<FzCheckboxProps, "size">;
