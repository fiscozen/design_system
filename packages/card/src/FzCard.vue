<template>
  <section :class="[sectionStaticClass, backgroundColor]">
    <header :class="[headerStaticClass, borderColor]">
      <h2 class="text-core-black text-2xl m-0 p-0">{{ props.title }}</h2>
    </header>
    <article :class="['p-20 flex-1', contentClass]">
      <slot></slot>
    </article>
    <footer v-if="atLeastOneButton" :class="[footerStaticClass, borderColor]">
      <FzIconButton
        v-if="tertiaryAction"
        @click="$emit('click:tertiary')"
        :iconName="tertiaryAction.icon"
        variant="invisible"
      />
      <FzButton
        v-if="secondaryAction"
        @click="$emit('click:secondary')"
        :label="secondaryAction.label"
        variant="secondary"
      />
      <FzButton
        v-if="primaryAction"
        @click="$emit('click:primary')"
        :label="primaryAction.label"
        variant="primary"
      />
    </footer>
  </section>
</template>

<script setup lang="ts">
import { computed, onMounted } from "vue";
import { FzCardColor, FzCardProps } from "./types";
import { FzButton, FzIconButton } from "@fiscozen/button";

const props = defineProps<FzCardProps>();
const sectionStaticClass =
  "border-1 border-solid border-grey-100 rounded flex flex-col w-fit";
const headerStaticClass = "h-72 border-b-1 border-solid p-20";
const footerStaticClass =
  "h-64 border-t-1 border-solid p-20 flex justify-end gap-8 items-center";

const backgroundColor = computed(() => {
  switch (props.color) {
    case FzCardColor.Blue:
      return "bg-background-alice-blue";
    case FzCardColor.Orange:
      return "bg-background-seashell";
    case FzCardColor.Purple:
      return "bg-background-pale-purple";
    default:
      return "bg-core-white";
  }
});

const borderColor = computed(() => {
  switch (props.color) {
    case FzCardColor.Blue:
      return "border-blue-200";
    case FzCardColor.Orange:
      return "border-orange-200";
    case FzCardColor.Purple:
      return "border-purple-200";
    default:
      return "border-grey-100";
  }
});

const atLeastOneButton = computed(
  () =>
    props.primaryAction !== undefined ||
    props.secondaryAction !== undefined ||
    props.tertiaryAction !== undefined,
);

onMounted(() => {
  if (props.tertiaryAction && !props.secondaryAction && !props.primaryAction)
    console.warn(
      "[Fiscozen Design System]: You should set primaryAction and secondaryAction if you want to set tertiaryAction",
    );
  else if (props.secondaryAction && !props.primaryAction)
    console.warn(
      "[Fiscozen Design System]: You should set primaryAction if you want to set secondaryAction",
    );
});
</script>
