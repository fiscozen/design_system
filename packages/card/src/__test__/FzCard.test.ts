import { mount } from "@vue/test-utils";
import { describe, it, expect, vi } from "vitest";
import FzCard from "../FzCard.vue";
import { FzCardColor } from "../types";

describe("FzCard", () => {
  it("renders correctly", async () => {
    const wrapper = mount(FzCard, {
      props: {
        title: "Test Card",
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toContain("Test Card");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("computes correct background color", async () => {
    const wrapper = mount(FzCard, {
      props: {
        title: "Test Card",
        color: FzCardColor.Orange,
      },
    });

    await wrapper.vm.$nextTick();
    expect(wrapper.find("section").classes()).toContain(
      "bg-background-seashell",
    );
  });

  it("computes correct border color", async () => {
    const wrapper = mount(FzCard, {
      props: {
        title: "Test Card",
        color: FzCardColor.Blue,
        primaryAction: {
          label: "Action 1",
          callback: () => {},
        },
      },
    });

    await wrapper.vm.$nextTick();
    expect(wrapper.find("header").classes()).toContain("border-blue-200");
    expect(wrapper.find("section").classes()).toContain("border-grey-100");
    expect(wrapper.find("footer").classes()).toContain("border-blue-200");
  });

  it("computes atLeastOneButton correctly", async () => {
    const wrapper = mount(FzCard, {
      props: {
        title: "Test Card",
        primaryAction: {
          label: "Action 1",
          callback: () => {},
        },
      },
    });

    await wrapper.vm.$nextTick();
    // @ts-ignore computed property are not typed
    expect(wrapper.vm.atLeastOneButton).toBe(true);
  });

  it("warns if tertiaryAction is set without primaryAction and secondaryAction", () => {
    const warnSpy = vi.spyOn(console, "warn").mockImplementation(() => {});
    mount(FzCard, {
      props: {
        title: "Test Card",
        tertiaryAction: {
          icon: "bell",
          callback: () => {},
        },
      },
    });
    expect(warnSpy).toHaveBeenCalledWith(
      "[Fiscozen Design System]: You should set primaryAction and secondaryAction if you want to set tertiaryAction",
    );
    warnSpy.mockRestore();
  });

  it("warns if secondaryAction is set without primaryAction", () => {
    const warnSpy = vi.spyOn(console, "warn").mockImplementation(() => {});
    mount(FzCard, {
      props: {
        title: "Test Card",
        secondaryAction: {
          label: "Action 2",
          callback: () => {},
        },
      },
    });

    expect(warnSpy).toHaveBeenCalledWith(
      "[Fiscozen Design System]: You should set primaryAction if you want to set secondaryAction",
    );
    warnSpy.mockRestore();
  });
});
