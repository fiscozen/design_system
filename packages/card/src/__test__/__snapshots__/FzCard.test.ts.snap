// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FzCard > renders correctly 1`] = `
"<section class="border-1 border-solid border-grey-100 rounded flex flex-col w-fit bg-core-white">
  <header class="h-72 border-b-1 border-solid p-20 border-grey-100">
    <h2 class="text-core-black text-2xl m-0 p-0">Test Card</h2>
  </header>
  <article class="p-20 flex-1"></article>
  <!--v-if-->
</section>"
`;
